php/index.php

```diff
@@ -1,45 +1,215 @@
-$data = array('code'         => $code,
-                'grant_type'   => 'authorization_code',
-                'redirect_uri' => $client_redirect);
-
-// use client_secret as basic auth password
-curl_setopt($ch, CURLOPT_USERPWD, $client_id . ':' . $client_secret);
-curl_setopt($ch, CURLOPT_RETURNTRANSFER,1);
-curl_setopt($ch, CURLOPT_POST, 1);
-curl_setopt($ch, CURLOPT_POSTFIELDS, $data);
-$response = json_decode(curl_exec($ch));
-
-$access_token = $response->access_token;
-echo("<p>Access token is " . $access_token . ".</p>");
-echo("<p>Using token to look up user information.</p>");
-
-$ch = curl_init($_ENV["CLEVER_API_BASE"] . "/me");
-// Use access token from previous call as bearer token to authenticate
-$header = array('Authorization: Bearer ' . $access_token);
-curl_setopt($ch, CURLOPT_HTTPHEADER, $header);
-curl_setopt($ch, CURLOPT_RETURNTRANSFER,1);
-$response = json_decode(curl_exec($ch));
-
-echo("<p>Here's some information about the user:</p>");
-echo("<pre>");
-print_r ($response);
-echo("</pre>");
-echo("<p>// Next: Look this user up in my database</p>");
-echo("<p>// Next: Create a session for them (or show an error message)</p>");
-
-?>
+/**
+ * This script demonstrates Clever Instant Login, using OAuth 2.0 for token acquistion & request authentication.
+ */
+
+error_reporting(E_ALL);
+
+// Handle incoming requests if we're running in a web server context
+if($_SERVER && array_key_exists('REQUEST_URI', $_SERVER)) {
+  process_incoming_requests($_SERVER['REQUEST_URI'], set_options());
+}
+
+/**
+ * Prepares options common to interacting with Clever's authentication & API
+ *
+ * @param   array $override_options  Options to oveverride from defaults
+ * @throws  Exception if configuration options are not adequately met
+ * @return  array $results           Options for use in Clever API requests
+ */
+function set_options(array $override_options = NULL) {
+  $options = array(
+    # Obtain your Client ID and secret from your Clever developer dashboard at https://account.clever.com/partner/applications
+    'client_id' => getenv('CLEVER_CLIENT_ID'),
+    'client_secret' => getenv('CLEVER_CLIENT_SECRET'),
+    'port' => getenv('CLEVER_PORT'),
+    'district_id' => getenv('CLEVER_DISTRICT_ID'),
+    'clever_oauth_base' => 'https://clever.com/oauth',
+    'clever_api_base' => 'https://api.clever.com',
+  );
+  if (isset($override_options)) {
+    array_merge($options, $override_options);
+  }
+
+  $options['clever_oauth_tokens_url'] = $options['clever_oauth_base'] . "/tokens";
+  $options['clever_oauth_authorize_url'] = $options['clever_oauth_base'] . "/authorize";
+  $options['clever_api_me_url'] = $options['clever_api_base'] . '/me';
+
+  // Clever redirect URIs must be preregistered on your developer dashboard.
+  // If using the default PORT set above, make sure to register "http://localhost:2587/oauth"
+  $options['client_redirect_url'] = "http://localhost:{$options['port']}/oauth";
```

Comment:
Another thought:  Because this is very procedural code that is written to demonstrate a process, maybe its worth doing an overview of the steps in the readme, linked to the steps in the code? 

1. Configure our server to receive oauth callbacks
2. Present the user with our "Login with Clever" button
3. User Clicks and is redirected to Clever for login
4. User is redirected back to our Application with a token that allows them to login
5. Application uses the token while communicating directly w/ clever to learn about the user
6. Application Allows/Denies access to based on what they learned from clever (eg, valid user?)